### PR TITLE
Implemented clean up old saves button

### LINKED ISSUE #1417: Implement clean up old saves button
"Clean Up Old Saves" button in the load menu is currently always disabled.
It should do the following: delete all except the latest quick save and auto save. It shouldn't touch the manual save types.
It should have a confirmation popup.

- [x] the button is enabled when there is at least 1 save file
- [x] logic for it, including a confirmation dialog. should delete all quick saves and autosaves except the last one.

### PR DESCRIPTION
**Pull request by DTanxxx**
Implemented the button using discussed logic. Modified SaveHelper.CountSaves to make counting numbers of auto and quick save files easier, for user information.

closes #1417

### CODE REVIEW

File: src/saving/SaveHelper.cs [hunk 1]
```diff
@@ -193,7 +193,7 @@ public static List<string> CreateListOfSaves(SaveOrder order = SaveOrder.LastMod
     /// <summary>
     ///   Counts the total number of saves and how many bytes they take up
     /// </summary>
-    public static (int count, long diskSpace) CountSaves()
+    public static (int count, long diskSpace) CountSaves(string nameStartsWith = "")
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:196]: I'd prefer to have the default value null

File: src/saving/SaveHelper.cs [hunk 2]
```diff
@@ -202,9 +202,12 @@ public static (int count, long diskSpace) CountSaves()
         {
             foreach (var save in CreateListOfSaves())
             {
-                file.Open(PathUtils.Join(Constants.SAVE_FOLDER, save), File.ModeFlags.Read);
-                ++count;
-                totalSize += file.GetLen();
+                if (save.StartsWith(nameStartsWith, StringComparison.CurrentCulture))
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:205]: and then here have the condition `nameStartsWith == null || ...`

File: src/saving/SaveHelper.cs [hunk 3]
```diff
@@ -242,6 +245,30 @@ public static void DeleteExcessSaves(string nameStartsWith, int maximumCount)
         }
     }
 
+    public static List<string> CleanUpOldSaves(string nameStartsWith)
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:248]: This method should be documented.

Comment (hhyyrylainen) [reply src/saving/SaveHelper.cs:248]: And if you keep this having the nameStartsWith parameter, this should be renamed to `CleanUpOldSavesOfType`

File: src/saving/SaveManagerGUI.cs [hunk 4]
```diff
@@ -107,6 +110,7 @@ public override void _Process(float delta)
             Math.Round((float)info.diskSpace / Constants.MEBIBYTE, 2).ToString(CultureInfo.CurrentCulture) + " MiB";
 
         UpdateSelectedCount();
+        UpdateButtonsStatus();
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:113]: I think that counting the saves for the cleanup old button should happen here, so that the filesystem doesn't get constantly scanned.

File: src/saving/SaveManagerGUI.cs [hunk 5]
```diff
@@ -148,6 +152,8 @@ private void UpdateButtonsStatus()
     {
         loadButton.Disabled = Selected.Count != 1;
         deleteSelectedButton.Disabled = Selected.Count == 0;
+        deleteOldButton.Disabled =
+            (SaveHelper.CountSaves("auto_save").count <= 1) && (SaveHelper.CountSaves("quick_save").count <= 1);
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:156]: So these counts should be cached.

File: src/saving/SaveManagerGUI.cs [hunk 6]
```diff
@@ -166,17 +172,39 @@ private void DeleteSelectedButtonPressed()
         deleteSelectedConfirmDialog.PopupCenteredMinsize();
     }
 
+    private void DeleteOldButtonPressed()
+    {
+        deleteOldConfirmDialog.DialogText =
+            "Deleting all old Auto and Quick saves cannot be undone, " +
+            "are you sure you want to permanently delete the following?\n" +
+            $" >>> {Mathf.Clamp(SaveHelper.CountSaves("auto_save").count - 1, 0, Settings.Instance.MaxAutoSaves)} " +
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:180]: I think it's fine to get the fresh counts to make sure that up to date info is shown to the user.
This math should be outside the text formatting. Also isn't Mathf always with floats? Whereas Math.Clamp(int, int) would be a method working on ints.

File: src/saving/SaveManagerGUI.cs [hunk 7]
```diff
@@ -166,17 +172,39 @@ private void DeleteSelectedButtonPressed()
...
+            "Quick save(s)";
+        deleteOldConfirmDialog.PopupCenteredMinsize();
+    }
+
     private void OnConfirmDeleteSelected()
     {
         GD.Print("Deleting save(s): ", string.Join(", ", Selected.Select(item => item.SaveName).ToList()));
 
         Selected.ForEach(item => SaveHelper.DeleteSave(item.SaveName));
-        deleteSelectedButton.Disabled = true;
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:174]: Why was this line removed?

File: src/saving/SaveManagerGUI.cs [hunk 8]
```diff
@@ -166,17 +172,39 @@ private void DeleteSelectedButtonPressed()
...
         RefreshList();
     }
 
+    private void OnConfirmDeleteOld()
+    {
+        string autoSaveNames = string.Join(", ", SaveHelper.CleanUpOldSaves("auto_save"));
+        string delimiter = string.Concat(Enumerable.Repeat(", ", Mathf.Clamp(autoSaveNames.Length, 0, 1)));
+        string quickSaveNames = string.Join(", ", SaveHelper.CleanUpOldSaves("quick_save"));
+
+        GD.Print("Deleting save(s): ", autoSaveNames, delimiter, quickSaveNames);
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:203]: This should be `Deleted` as they are already deleted at this point.

File: src/saving/SaveManagerGUI.cs [hunk 9]
```diff
@@ -166,17 +172,39 @@ private void DeleteSelectedButtonPressed()
...
-        deleteSelectedButton.Disabled = true;
         selected = null;
 
         RefreshList();
     }
 
+    private void OnConfirmDeleteOld()
+    {
+        string autoSaveNames = string.Join(", ", SaveHelper.CleanUpOldSaves("auto_save"));
+        string delimiter = string.Concat(Enumerable.Repeat(", ", Mathf.Clamp(autoSaveNames.Length, 0, 1)));
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:200]: What does this do? If you are checking for if anything was added, then just use an if and check if autoSaveNames is length of < 1

File: src/saving/SaveManagerGUI.tscn [hunk 10]
```diff
@@ -142,9 +143,22 @@ size_flags_vertical = 5
...
 dialog_text = "Text not set..."
+
+[node name="DeleteOldConfirmDialog" type="ConfirmationDialog" parent="CenterContainer/VBoxContainer"]
+margin_right = 184.0
+margin_bottom = 97.0
+rect_min_size = Vector2( 0, 0 )
+size_flags_horizontal = 5
+size_flags_vertical = 5
+popup_exclusive = true
+window_title = "Delete old Saves"
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.tscn:154]: This should probably end in a question mark as the popup is a confirmation on deleting them.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/saving/SaveManagerGUI.cs (hunk 7 above)
Comment (DTanxxx) [reply src/saving/SaveManagerGUI.cs:174]: It was removed because since UpdateButtonsStatus was moved to _Process, after every RefreshList call the buttons are automatically updated and there is no need to disable it before the RefreshList call.

Comment (hhyyrylainen) [reply src/saving/SaveManagerGUI.cs:174]: But that happens asynchronously, so if it takes a long time to refresh the list, it would be good to disable the button until the list is refreshed.

File: src/saving/SaveManagerGUI.cs (hunk 9 above)
Comment (DTanxxx) [reply src/saving/SaveManagerGUI.cs:200]: Yes, it either gives a comma separating auto save names and quick save names, or gives empty string when there are no auto save names. I added this line to reduce the line count since if statement would take more lines. I can change it to an if statement though.

Comment (hhyyrylainen) [reply src/saving/SaveManagerGUI.cs:200]: I would prefer the if for readablility reason as it took a minute for me to figure out what this code does, whereas conditionally adding a comma is very quick to understand when reading through it the first time.

File: src/saving/SaveManagerGUI.cs (hunk 7 above)
Comment (DTanxxx) [reply src/saving/SaveManagerGUI.cs:174]: Ah I see, got it.

File: src/saving/SaveManagerGUI.cs (hunk 9 above)
Comment (DTanxxx) [reply src/saving/SaveManagerGUI.cs:200]: Ok, if it is for readability then it is pretty annoying to read indeed.

File: src/saving/SaveManagerGUI.cs (hunk 6 above)
Comment (DTanxxx) [reply src/saving/SaveManagerGUI.cs:180]: I tried to replace Mathf.Clamp with Math.Clamp but it turns out Math does not have that member function. I looked up Microsoft docs and it says the Math.Clamp API is non CLS-Compliant. Does that have something to do with Visual Studio not recognising the API?

Comment (hhyyrylainen) [reply src/saving/SaveManagerGUI.cs:180]: Actually it looks like we have our own Clamp: https://github.com/Revolutionary-Games/Thrive/blob/0e79166eedcc408980e2215b63cba2c37acb926b/src/general/MathUtils.cs#L12
Which should work like `(SaveHelper.CountSaves("auto_save").count - 1).Clamp(0, Settings.Instance.MaxAutoSaves)`

Comment (DTanxxx) [reply src/saving/SaveManagerGUI.cs:180]: Ah I see, thanks.

File: project.godot [hunk 11]
```diff
@@ -154,13 +154,13 @@ e_rotate_left={
 }
 e_undo={
 "deadzone": 0.5,
-"events": [ Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"alt":false,"shift":false,"control":true,"meta":false,"command":true,"pressed":false,"scancode":90,"unicode":0,"echo":false,"script":null)
+"events": [ Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"alt":false,"shift":false,"control":true,"meta":true,"command":true,"pressed":false,"scancode":90,"unicode":0,"echo":false,"script":null)
```
Comment (hhyyrylainen) [inline project.godot:157]: Changes in this file are probably unintentional...

File: src/saving/SaveManagerGUI.cs [hunk 12]
```diff
@@ -166,6 +177,19 @@ private void DeleteSelectedButtonPressed()
...
 
+    private void DeleteOldButtonPressed()
+    {
+        int autoSavesToDeleteCount = (currentAutoSaveCount - 1).Clamp(0, Settings.Instance.MaxAutoSaves);
+        int quickSavesToDeleteCount = (currentQuickSaveCount - 1).Clamp(0, Settings.Instance.MaxQuickSaves);
+
+        deleteOldConfirmDialog.DialogText =
+            "Deleting all old Auto and Quick saves cannot be undone, " +
+            "are you sure you want to permanently delete the following?\n" +
+            $" >>> {autoSavesToDeleteCount} Auto save(s)\n" +
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:188]: I think that using ` - ` instead of ` >>> ` would look more consistent with the rest of the game.

File: src/saving/SaveManagerGUI.cs [hunk 13]
```diff
@@ -106,7 +112,11 @@ public override void _Process(float delta)
         totalSaveSize.Text =
             Math.Round((float)info.diskSpace / Constants.MEBIBYTE, 2).ToString(CultureInfo.CurrentCulture) + " MiB";
 
+        currentAutoSaveCount = SaveHelper.CountSaves("auto_save").count;
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:115]: Have you tested if this is too slow with like 100 save files?
If this causes lag then it would feel better if this was computed in a background thread.

File: src/saving/SaveManagerGUI.cs [hunk 14]
```diff
@@ -177,6 +201,22 @@ private void OnConfirmDeleteSelected()
         RefreshList();
     }
 
+    private void OnConfirmDeleteOld()
+    {
+        string autoSaveNames = string.Join(", ", SaveHelper.CleanUpOldSavesOfType("auto_save"));
```
Comment (hhyyrylainen) [inline src/saving/SaveManagerGUI.cs:206]: I think you could use just a single string, or maybe a stringbuilder if a string creation is too slow.
Very rough code:
```
string message = string.Join(", ", SaveHelper.CleanUpOldSavesOfType("auto_save"));
if(message.Length > 0)
    message += ", ";
message += string.Join(", ", SaveHelper.CleanUpOldSavesOfType("quick_save"));
```

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

**Review by hhyyrylainen — APPROVED** [review]
